### DISCUSSION TITLE
TSC: Improving Community Experience (May 2021)

### DISCUSSION THREAD

**Discussion by hcgatewood** [magma/magma#6813] [Communications]
# Improving Community Experience

***Magma community experience was unpleasant and under-productive. We've been working to improve it.***

## Context

At the formation of the TSC, we identified two key areas where the community was experiencing pain points

* 🚨 *Onboarding community members*, both to the community and to the dev process
* 🚨 *Communicating developer expectations*

To that end, we identified and landed updates to the community experience along 3 axes

* 💧 *Streamline* dev process
* 📝 *Document* dev process + expectations
* 🎙 *Publicize* project resources

### Pain points

Getting a PR merged requires two things: *maintainer approvals* and *passing CI checks*. Garnering sufficient approvals and passing required CI checks was a non-trivial task which authors were often unable to navigate.

* PR authors didn't know:
    * Required maintainer approvals
    * Who to message when blocked (review or failing CI)
* Maintainers didn't know:
    * Which PRs needed their attention
    * Which maintainer should review a PR
    * When a PR was ready to be merged
* Manual review required to upkeep PR health

### Resolutions

Review process

* ✅ [Team-based GitHub codeowners](https://github.com/magma/magma/discussions/6218) (`approvers-*` *teams)*
    * Authors understand required maintainer approvals
    * Maintainers know which maintainer should review a PR
    * Related: [Clarified project permissions](https://github.com/magma/magma/discussions/5784) (and [Permissions changelist](https://github.com/magma/magma/discussions/5866)), [Clarified project teams](https://github.com/magma/magma/discussions/6301)
* ✅ [Toast](https://magmacore.slack.com/apps/AC6M092SD-toast) Slack integration
    * Maintainers are automatically messaged when action required
* ✅ [`ready2merge` label](https://github.com/magma/magma/labels/ready2merge)
    * Maintainers know when a PR is ready to be merged
* ✅ [CI checks documentation](https://magma.github.io/magma/docs/next/contributing/contribute_ci_checks)
    * Authors know who to message when blocked on failing CI

PR health

* ✅ Automate linting and static analysis
    * `cloud_lint` style and static analysis for Orc8r
    * [reviewdog](https://github.com/reviewdog/reviewdog)-managed linters for: GCC warnings, Python, shell, Dockerfiles, YAML, and common misspellings
    * [CodeQL](https://securitylab.github.com/tools/codeql/) static analysis for: Go, Python, JavaScript
* ✅ [Codecov](https://app.codecov.io/gh/magma/magma) for full-project code coverage metrics
    * Track code coverage over time, identify under-covered parts of codebase
    * Inline code coverage directly in pull requests (via [Sourcegraph](https://docs.sourcegraph.com/integration/browser_extension) and its [Codecov extension](https://sourcegraph.com/extensions/sourcegraph/codecov))

<img width="1792" alt="Screen Shot 2021-05-05 at 4 28 43 PM" src="https://user-images.githubusercontent.com/8029544/117363056-abb96400-ae81-11eb-9b15-36f81c970845.png">

CI health

* ✅ Pull critical tests into merge-blocking CI checks
    * `session_manager_test` unit test *session_manager*
    * `mme_test` unit test *sctpd* and *mme*
    * `cwag-build` build CWAG containers
    * `feg-build` added a check for container crashloop
* ✅ [Process for freezing or reverting breaking PRs](https://github.com/magma/magma/issues/6801)
    * “CI oncall” to manually manage post-merge CI checks, to determine whether master is healthy
    * Guidance for CI oncall for reverting breaking PRs
    * Guidance for CI oncall for freezing master branch when master is unhealthy

### Pain points

Contributing to the Magma project is a non-trivial endeavor — it's a project with multiple languages, platforms, and applications, and requires hardware in the validation path.

Community contributors struggled to understand baseline development workflow and expectations around contributions.

### Resolutions

General

* ✅ Reorganize [Docusaurus](https://docs.magmacore.org/docs/basics/introduction.html) sidebar with community focus, aligned to "what do I want to do"

Onboarding

* ✅ [Developer Onboarding](https://magma.github.io/magma/docs/next/contributing/contribute_onboarding) onboards developers completely new to Magma

Contributing

* ✅ [Development Workflow](https://magma.github.io/magma/docs/next/contributing/contribute_workflow) covers required PR practices, including an opinionated description of best-practices
* ✅ [Contributing Conventions](https://magma.github.io/magma/docs/next/contributing/contribute_conventions) covers conventions and expectations for code contributions — for the project as a whole, documentation, specific languages (Go, Python, and C++ forthcoming), specific tools (gRPC, Swagger, YAML), and more
* ✅ [Continuous Integration Checks](https://magma.github.io/magma/docs/next/contributing/contribute_ci_checks) annotates CI checks with purpose and points of contact

Maintaining

* ✅ [Codeowners Workflow](https://magma.github.io/magma/docs/next/contributing/contribute_codeowners) provides guidance to codeowners on shepherding community contributions and when force-merging PRs is acceptable
* ✅ [Proposals Process](https://magma.github.io/magma/docs/next/proposals/p014_proposal_process) provides guidance on proposing new features, getting feedback from the community, and aligning on a final design

### Pain points

While community-facing documentation now exists, it's still siloed in the Magma Docusaurus — which has 130 readme files at the time of this writing.

New community members don't know to sift through the documentation. Instead, we can bring the documentation to them at key touch points.

### Resolutions

* ✅ [Community calendar](https://calendar.google.com/calendar/u/0/embed?src=c_gbiu1t7a67ika1th2smldeh19s@group.calendar.google.com) for inviting entire community to public meetings
* ✅ [Revamp project README](https://github.com/magma/magma/blob/master/README.md) with community entry points, contributing resources, and a modernized hero banner

<img width="1015" alt="Screen Shot 2021-05-05 at 5 20 05 PM" src="https://user-images.githubusercontent.com/8029544/117363151-ce4b7d00-ae81-11eb-9e50-a3603313928b.png">

* ✅ [GreetBot](https://magmacore.slack.com/apps/A2LJCT3GR-greetbot) Slack integration to welcome new Slack members with community resources

<img width="828" alt="Screen Shot 2021-05-03 at 11 10 29 PM" src="https://user-images.githubusercontent.com/8029544/117363188-dd322f80-ae81-11eb-9fbd-02f59db2e93f.png">

* ✅ [PR bot](https://github.com/magma/magma/blob/9cb7401509af2a03a349196192ec6e3e460f6a57/.github/workflows/pr_bot.yml) to comment on opened pull requests, welcoming authors with how to get PR merged and follow community standards

<img width="1015" alt="Screen Shot 2021-05-06 at 9 50 18 AM" src="https://user-images.githubusercontent.com/8029544/117363213-e3281080-ae81-11eb-8f3f-8f2f46522fa6.png">

## Conclusion

We're still looking to do more. Please follow up in comments on this post with your feedback, and ways we can improve the process of *onboarding community members* and *communicating developer expectations*.

Huge #thanks to @themarwhal and @electronjoe for their fantastic work in driving this effort forward. And big #thanks to @PARitter and Gale for advising and supporting the improvements!

And #thanksteam for pulling together to support the Magma community 💯🌟

**Comment by themarwhal** [comment]
Thanks @hcgatewood for being a fantastic TSC chair and for this great summary. 🙏
Excited to get more things done!

![](https://media4.giphy.com/media/IwAZ6dvvvaTtdI8SD5/giphy.gif)

**Comment by shawaj** [comment]
the [slack community invite](https://magmacore.slack.com/signup#/domain-signup) does not work unless you have an email address from one of the following domains:

- openstack.org
- freedomfi.com
- openairinterface.org
- wavelabs.ai
- altencalsoftlabs.com

The documentation implies this is an open community - is this an oversight or is it intentional?

**Reply by amarpad** [reply]
Must be an oversight. cc @galemccommons, not sure who has the keys to the slack account.

**Reply by wendallkaters** [reply]
Thank you for pointing out this out! Not sure why the settings changed. Updating the redirect link. Correct link should be associated with slack.magmacore.org now.

**Reply by ulaskozat** [reply]
It still does not work, I think.

**Reply by PARitter** [reply]
The invite link "https://slack.magmacore.org" as shared by @wendallkaters has been tested and confirmed to work.  The original link referenced above "https://magmacore.slack.com/signup#/domain-signup" does appear to still be broken.

The website is being updated to use the link provided by Kendall.  We should update the rest of our collateral to use it too.
